pkg/mutation/base_mutator.go

```diff
@@ -0,0 +1,122 @@
+package mutation
+
+import (
+	"fmt"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/path/parser"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func NewMutator(path parser.Path, value string) BaseMutator {
+	return BaseMutator{
+		path:  path,
+		Value: value,
+	}
+}
+
+// Base is a base mutator wrapping the Assign and AssignMetadata type.
+type BaseMutator struct {
+	path  parser.Path
+	Value string
+}
+
+func (m BaseMutator) Mutate(obj *unstructured.Unstructured) error {
+	return m.mutate(obj.Object, nil, 0)
+}
+
+func (m BaseMutator) mutate(current interface{}, previous interface{}, depth int) error {
+	pathEntry := m.Path().Nodes[depth]
+	if len(m.Path().Nodes)-1 == depth {
+		current.(map[string]interface{})[pathEntry.(parser.Object).Reference] = m.Value
```

Comment:
I added the code to handle the last element being a list element.
So far we can only accept string values for elements, so to add a new container we would have to do two steps:
Rule 1: spec.containers[name: "sidecar"]
this will add a new element to the list, with name=sidecar
Rule 2: spec.containers[name: "sidecar"].image  value="foo"
which would add image=foo to the previously created item

I suppose we could think about adding complex values to do this in one step later (another PR)